presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp

```diff
@@ -66,43 +68,125 @@ std::unique_ptr<BroadcastFileInfo> BroadcastFileInfo::deserialize(
 }
 
 BroadcastFileWriter::BroadcastFileWriter(
-    std::string_view filename,
+    const std::string& filename,
     const RowTypePtr& inputType,
-    std::shared_ptr<velox::filesystems::FileSystem> fileSystem,
+    uint64_t maxPageSize,
     velox::memory::MemoryPool* pool)
-    : fileSystem_(std::move(fileSystem)),
-      filename_(filename),
-      numRows_(0),
-      maxSerializedSize_(0),
+    : inputType_(inputType),
+      maxPageSize_(maxPageSize),
       pool_(pool),
-      serde_(std::make_unique<serializer::presto::PrestoVectorSerde>()),
-      inputType_(inputType) {}
+      currentStreamArena_(std::make_unique<velox::StreamArena>(pool_)),
+      serde_(std::make_unique<serializer::presto::PrestoVectorSerde>()) {
+  writeFile_ = velox::filesystems::getFileSystem(filename, nullptr)
+                   ->openFileForWrite(filename);
+}
+
+void BroadcastFileWriter::maybeInitializeSerializer() {
+  if (currentPageSerializer_ != nullptr) {
+    return;
+  }
+  currentPageSerializer_ = serde_->createIterativeSerializer(
+      inputType_, 1, currentStreamArena_.get());
+}
+
+void BroadcastFileWriter::flushCurrentPage() {
+  VELOX_CHECK_NOT_NULL(currentPageSerializer_);
+
+  IOBufOutputStream pageStream(*pool_);
+  currentPageSerializer_->flush(&pageStream);
+
+  auto serializedData = pageStream.getIOBuf();
+  VELOX_CHECK_NOT_NULL(serializedData);
+
+  const auto pageSize = serializedData->computeChainDataLength();
+  VELOX_CHECK_GT(pageSize, 0);
+
+  for (auto& buf : *serializedData) {
+    writeFile_->append(std::string_view(
+        reinterpret_cast<const char*>(buf.data()), buf.size()));
+  }
+
+  writeFile_->flush();
+  totalFileSize_ += pageSize;
+
+  // Track page size for footer
+  pageSizes_.push_back(pageSize);
+
+  // Clear the current page serializer to prepare for next page.
+  currentPageSerializer_->clear();
+}
+
+void BroadcastFileWriter::write(const RowVectorPtr& rowVector) {
+  if (rowVector->size() == 0) {
+    return;
+  }
+
+  maybeInitializeSerializer();
```

Comment:
Can we pass numRows to maybeInitializeSerializer() for the initial rows on the first call?